### ISSUE TITLE
Speed Diff: mx.mod.Module vs. mx.model.FeedForward

### ISSUE THREAD

**Issue by ilkarman** [apache/incubator-mxnet#3151]
**Edit: To give a quick summary -> is the mx.mod.Module() method slower than the usual ctx.bind(.) etc which is called with mx.model.FeedForward?**

**Using mx.model.FeedForward I get 240 a second, with mx.mod.Module I get 160 a second.**

I can't quite figure out why, but method A is a bit faster (30% faster maybe):

**Method A ("standard"):**

Where I have train_x, train_y loaded into RAM (140GB). So I load a raw data-set, generate my features (this takes maybe 30 minutes) and then keep in RAM:

```
train_iter = mx.io.NDArrayIter(train_x, train_y, batch_size=BATCH_SIZE, shuffle=True)
model = mx.model.FeedForward(
    ...
)
model.fit(
    ...
)
```

**Method B:**

I keep the raw data-set in RAM (4GB), however I create features on the fly using this function:

```
def load_data_frame(X_data, y_data, batch_size=128, shuffle=False):
    """
    For low RAM this methods allows us to keep only the original data
    in RAM and calculate the features (which are orders of magnitude bigger
    on the fly). This keeps only 10 batches worth of features in RAM using
    asynchronous programing and yields one DataBatch() at a time.
    """

    if shuffle:
        idx = X_data.index
        assert len(idx) == len(y_data)
        rnd = np.random.permutation(idx)
        X_data = X_data.reindex(rnd)
        y_data = y_data[rnd]

    # Dictionary to create character vectors
    character_hash = pd.DataFrame(np.identity(len(ALPHABET), dtype='bool'), columns=ALPHABET)

    # Yield processed batches asynchronously
    # Buffy 'batches' at a time
    def async_prefetch_wrp(iterable, buffy=10):
        poison_pill = object()

        def worker(q, it):
            for item in it:
                q.put(item)
            q.put(poison_pill)

        queue = Queue.Queue(buffy)
        it = iter(iterable)
        thread = threading.Thread(target=worker, args=(queue, it))
        thread.daemon = True
        thread.start()
        while True:
            item = queue.get()
            if item == poison_pill:
                return
            else:
                yield item

    # Async wrapper around
    def async_prefetch(func):
        @functools.wraps(func)
        def wrapper(*args, **kwds):
            return async_prefetch_wrp(func(*args, **kwds))
        return wrapper

    @async_prefetch
    def feature_extractor(dta, val):
        # Yield mini-batch amount of character vectors
        X_split = np.zeros([batch_size, 1, FEATURE_LEN, len(ALPHABET)], dtype='bool')
        for ti, tx in enumerate(dta):
            chars = list(tx)
            for ci, ch in enumerate(chars):
                if ch in ALPHABET:
                    X_split[ti % batch_size][0][ci] = np.array(character_hash[ch])
            # No padding -> only complete batches processed
            if (ti + 1) % batch_size == 0:
                yield mx.nd.array(X_split), mx.nd.array(val[ti + 1 - batch_size:ti + 1])
                X_split = np.zeros([batch_size, 1, FEATURE_LEN, len(ALPHABET)], dtype='bool')

    # Yield one mini-batch at a time and asynchronously process to keep 4 in queue
    for Xsplit, ysplit in feature_extractor(X_data, y_data):
        yield DataBatch(data=[Xsplit], label=[ysplit])
```

And then I train like so:

```
# TRAINING (and Testing):
# Create mx.mod.Module()
cnn = create_crepe()
mod = mx.mod.Module(cnn, context=ctx)

# Bind shape
mod.bind(data_shapes=[('data', DATA_SHAPE)],
         label_shapes=[('softmax_label', (BATCH_SIZE,))])

# Initialise parameters and optimiser
mod.init_params(mx.init.Normal(sigma=SD))
mod.init_optimizer(optimizer='sgd',
                   optimizer_params={
                       "learning_rate": 0.01,
                       "momentum": 0.9,
                       "wd": 0.00001,
                       "rescale_grad": 1.0/BATCH_SIZE
                   })

# Load Data
X_train, y_train = load_file('amazon_review_polarity_train.csv')
X_test, y_test = load_file('amazon_review_polarity_test.csv')

# Evaluation metric:
metric = mx.metric.Accuracy()

# Train EPOCHS
for epoch in range(EPOCHS):
    t = 0
    metric.reset()
    tic_in = time.time()
    for batch in load_data_frame(X_data=X_train,
                                 y_data=y_train,
                                 batch_size=BATCH_SIZE,
                                 shuffle=True):
        # Push data forwards and update metric
        # For training + testing
        mod.forward(batch, is_train=True)
        mod.update_metric(metric, batch.label)
        # Get weights and update
        # For training only
        mod.backward()
        mod.update()
        # Log every 10 batches
        t += 1
        if t % 10 == 0:
            train_t = time.time() - tic_in
            metric_m, metric_v = metric.get()
            print("epoch: %d iter: %d metric(%s): %.4f dur: %.0f" % (epoch, t, metric_m, metric_v, train_t))
```

The reason I like method B is because it gives me complete control over many aspects - it's very easy for me to limit this to 8GB total RAM (compared to 140GB for example).

Now I understand that holding all features in RAM would be fast, however:
- I asynchronously prefetch batches to keep a queue of 10; so after the first batch I should have no cost generating features on the fly because that is done multi-threaded whilst my GPU is processing the weights + gradients
- I store the data as dtype='bool' because that costs me 4 bits compared to say 32 bits and all my data is just a binary 1/0

**Comment by Pagey** [comment]
No Comments on this? I am getting a similar performance degradation (i.e. around 66%) on latest version (0.9.5) on a GTX 1080ti GPU on windows for  mx.mod.Module relative to  mx.model.FeedForward
The degradation can be visualized on GPU-Z where GPU Load graph is choppy and oscillates between 100% and 0% per batch for mx.mod.Module, and seems continuous around 100% for mx.model.FeedForward (see attached files)
![gpuz_feedfoward](https://cloud.githubusercontent.com/assets/28292920/25944746/1526605c-364d-11e7-9f3a-bbb670d06328.png)
![gpuz_module](https://cloud.githubusercontent.com/assets/28292920/25944747/152b0d82-364d-11e7-861e-60abbe5ef6a8.png)


the codes these are running are identical except for the model definition and training:

--FeedFoward version:

model = mx.model.FeedForward(
    ctx=devices,
    symbol = da_net,       
    num_epoch = 2000,    
    learning_rate = 0.00001,
    wd=1e-5,
    initializer = mx.init.Xavier(rnd_type='gaussian', factor_type="in", magnitude=2),
    optimizer='adam' 
)
model.fit(
    X=train,       # training data
    eval_data=val, # validation data
    eval_metric='acc',#.metric.Accuracy,
    batch_end_callback = batch_end_callbacks,
)

-- Module version:

model = mx.mod.Module(context = devices, symbol = da_net)
model.fit(train_data=train, eval_data =val ,eval_metric='acc', num_epoch=2000,batch_end_callback =batch_end_callbacks, optimizer='adam',optimizer_params={'learning_rate':0.00001,'wd':1e-5},initializer = mx.init.Xavier(rnd_type='gaussian', factor_type="in", magnitude=2))

**Comment by piiswrong** [comment]
@eric-haibin-lin 
Could you provide the whole script including data iterator and network def?

**Comment by Pagey** [comment]
it would take me some time to produce a minimal viable code, but generally the network is a variation of alexnet, and the data iterators are on arrays of images, i.e.

train = mx.io.NDArrayIter(
        train_img, train_lbl, batcher, shuffle=True)
val = mx.io.NDArrayIter(
        val_img, val_lbl, batcher)

**Comment by Pagey** [comment]
import mxnet as mx
import numpy as np
import matplotlib.pyplot as plt
from symbols import alexnet as nettt
import sys
import time

print ('started ' + time.strftime('%x %X'))
startedd = time.time()

target_dim = 256

batcher = 64

num_images=512

use_module_on_your_own_risk=True

# create images
# rand (0) rand with circles(1)

xcoords = np.ones((target_dim,1))*np.arange(target_dim)
ycoords = np.transpose(xcoords)

def add_random_circle(da_img,xcoords,ycoords):
    tolerance=5

    center_x = int(np.random.rand() * target_dim)
    center_y = int(np.random.rand() * target_dim)
    radius=np.random.rand() * target_dim / 2.0

    distances= np.power(np.power((xcoords-center_x),2)+np.power((ycoords-center_y),2),0.5)

    da_img[np.where(np.abs(distances-radius)<tolerance)] = 1

    return da_img


nn_train = num_images;
nn_val = num_images;

image_list = np.random.rand(nn_train*2,target_dim,target_dim)
llabel = np.round(np.random.rand(nn_train*2))

train_indeces = np.arange(nn_train)
val_indeces= nn_train + np.arange(nn_train)

train_lbl=llabel[train_indeces]
val_lbl=llabel[val_indeces]


train_img = np.zeros((nn_train,1,target_dim,target_dim))
for indexer in range(nn_train):
    img_now = image_list[train_indeces[indexer]]
    if train_lbl[indexer] >0:
        img_now=add_random_circle(img_now,xcoords,ycoords)

    train_img[indexer,0,:,:]=img_now.astype(np.float32)/255

nn_val=len(val_indeces)
val_img = np.zeros((nn_val,1,target_dim,target_dim))

for indexer in range(nn_val):
    img_now = image_list[val_indeces[indexer]]
    if val_lbl[indexer] > 0:
        img_now = add_random_circle(img_now, xcoords, ycoords)

    val_img[indexer,0,:,:]=img_now.astype(np.float32)/255

# normalizing by removing mean of training set!
meaner=np.mean(train_img)

print('mean train before: '+str(np.mean(train_img)))
print('mean val before: '+str(np.mean(val_img)))

train_img -= meaner
val_img -= meaner

train = mx.io.NDArrayIter(
        train_img, train_lbl, batcher, shuffle=True)
val = mx.io.NDArrayIter(
        val_img, val_lbl, batcher)

print(train_img.shape)
print(train_lbl.shape)
print(val_img.shape)
print(val_lbl.shape)

da_net = nettt.get_symbol(num_classes=2)

devices = mx.gpu()

import logging

logging.getLogger().setLevel(logging.DEBUG)

batch_end_callbacks = [mx.callback.log_train_metric(5)]


if not use_module_on_your_own_risk:

    model = mx.model.FeedForward(
        ctx=devices,
        symbol = da_net,       # network structure
        num_epoch = 100,     # number of  passes for training
        learning_rate = 0.00001,
        wd=1e-5,
        #initializer = mx.init.Normal(),
        initializer = mx.init.Xavier(rnd_type='gaussian', factor_type="in", magnitude=2),
        optimizer='adam'
    )
    model.fit(
        X=train,       # training data
        eval_data=val, # validation data
        eval_metric='acc',#.metric.Accuracy,
        batch_end_callback = batch_end_callbacks,
    )


else:
    model = mx.mod.Module(context = devices, symbol = da_net)

    ##train
    model.fit(train_data=train, eval_data =val ,eval_metric='acc', num_epoch=100,batch_end_callback =batch_end_callbacks, optimizer='adam',optimizer_params={'learning_rate':0.00001,'wd':1e-5},initializer = mx.init.Xavier(rnd_type='gaussian', factor_type="in", magnitude=2))

finishedd= time.time()
print ('finished ' + time.strftime('%x %X') )
print ('sex:' + str(finishedd - startedd) )

**Comment by Pagey** [comment]
ok about toy code demonstrates the problem, about *2 time difference on my GPU (and intermittent  GPU load on GPUZ)
Thanks

**Comment by piiswrong** [comment]
Could be a problem with NDArrayIter. Does this also happen with ImageRecordIter?

**Comment by Pagey** [comment]
I only worked with NDArrayIter so far as my images could fit in RAM-
could someone reproduce the result with my code?

**Comment by Pagey** [comment]
OK just checked that a similar time factor (this time around *1.8 time in Module compared with Feedforward) appears also when using ImageRecordIter
in a toy example, this time using LeNet and cifar10 rec files as prepared in the image-classification demo
this is the code:
`import mxnet as mx
import numpy as np
import matplotlib.pyplot as plt
from symbols import lenet as nettt
import sys
import time

print ('started ' + time.strftime('%x %X'))
startedd = time.time()

target_dim = 256

batcher = 64*8


epox=500;

use_module_on_your_own_risk=False#False#True

train= mx.io.ImageRecordIter(path_imgrec ='data\cifar10_train.rec',batch_size=batcher,data_shape=(3,28,28),shuffle=True)
val= mx.io.ImageRecordIter(path_imgrec ='data\cifar10_val.rec',batch_size=batcher,data_shape=(3,28,28))

da_net = nettt.get_symbol(num_classes=10)

devices = mx.gpu()

import logging

logging.getLogger().setLevel(logging.DEBUG)

batch_end_callbacks = [mx.callback.log_train_metric(50)]


if not use_module_on_your_own_risk:

    model = mx.model.FeedForward(
        ctx=devices,
        symbol = da_net,       # network structure
        num_epoch = epox,     # number of  passes for training
        learning_rate = 0.0001,
        wd=1e-5,
        #initializer = mx.init.Normal(),
        initializer = mx.init.Xavier(rnd_type='gaussian', factor_type="in", magnitude=2),
        optimizer='adam'
    )
    model.fit(
        X=train,       # training data
        eval_data=val, # validation data
        eval_metric='acc',#.metric.Accuracy,
        batch_end_callback = batch_end_callbacks,
    )


else:
    model = mx.mod.Module(context = devices, symbol = da_net)

    ##train
    model.fit(train_data=train, eval_data =val ,eval_metric='acc', num_epoch=epox,batch_end_callback =batch_end_callbacks, optimizer='adam',optimizer_params={'learning_rate':0.00001,'wd':1e-5},initializer = mx.init.Xavier(rnd_type='gaussian', factor_type="in", magnitude=2))

finishedd= time.time()
print ('finished ' + time.strftime('%x %X') )
print ('sex:' + str(finishedd - startedd) )`

**Comment by Pagey** [comment]
the above examples (mine) don't seem to show a difference between module and feedfoward speed in version 0.10.1 👍

**Comment by szha** [comment]
This issue is closed due to lack of activity in the last 90 days. Feel free to ping me to reopen if this is still an active issue. Thanks!
